tests/integration/cache_test.go

```diff
@@ -160,98 +199,160 @@ func testWatch(t *testing.T, kv clientv3.KV, watcher Watcher) {
+			wantEvents: []*clientv3.Event{rev11PutFooFuture, rev12DeleteFooFuture},
+		},
+		{
+			name:       "Watch from rev 12 with txn prefix /foo",
+			key:        "/foo",
+			opts:       []clientv3.OpOption{clientv3.WithRev(12), clientv3.WithPrefix()},
+			wantEvents: []*clientv3.Event{rev12PutFooTx1, rev12DeleteFooFuture, rev12PutFooTx2},
 		},
 	}
 
-	t.Log("Open test watchers")
-	watches := make([]clientv3.WatchChan, len(tcs))
-	for i, tc := range tcs {
-		watches[i] = watcher.Watch(ctx, tc.key, tc.opts...)
-	}
-	t.Log("Setup data")
-	if _, err := kv.Put(ctx, string(event1PutFooA.Kv.Key), string(event1PutFooA.Kv.Value)); err != nil {
+	t.Log("Write the first batch of events rev 2-10")
+	if _, err := kv.Put(ctx, string(rev2PutFooA.Kv.Key), string(rev2PutFooA.Kv.Value)); err != nil {
 		t.Fatalf("Put: %v", err)
 	}
-	if _, err := kv.Put(ctx, string(event2PutFooB.Kv.Key), string(event2PutFooB.Kv.Value)); err != nil {
+	if _, err := kv.Put(ctx, string(rev3PutFooB.Kv.Key), string(rev3PutFooB.Kv.Value)); err != nil {
 		t.Fatalf("Put: %v", err)
 	}
-	if _, err := kv.Delete(ctx, string(event3DeleteFooA.Kv.Key)); err != nil {
+	if _, err := kv.Delete(ctx, string(rev4DeleteFooA.Kv.Key)); err != nil {
 		t.Fatalf("Delete: %v", err)
 	}
-	if _, err := kv.Txn(ctx).Then(clientv3.OpPut(string(event4PutFooA.Kv.Key), string(event4PutFooA.Kv.Value)), clientv3.OpDelete(string(event5DeleteFooB.Kv.Key))).Commit(); err != nil {
+	if _, err := kv.Txn(ctx).Then(clientv3.OpPut(string(rev5PutFooA.Kv.Key), string(rev5PutFooA.Kv.Value)), clientv3.OpDelete(string(rev5DeleteFooB.Kv.Key))).Commit(); err != nil {
 		t.Fatalf("Txn: %v", err)
 	}
-	if _, err := kv.Put(ctx, string(event6PutFooC.Kv.Key), string(event6PutFooC.Kv.Value)); err != nil {
+	if _, err := kv.Put(ctx, string(rev6PutFooC.Kv.Key), string(rev6PutFooC.Kv.Value)); err != nil {
 		t.Fatalf("Put: %v", err)
 	}
-	if _, err := kv.Put(ctx, string(event7PutFooBar.Kv.Key), string(event7PutFooBar.Kv.Value)); err != nil {
+	if _, err := kv.Put(ctx, string(rev7PutFooBar.Kv.Key), string(rev7PutFooBar.Kv.Value)); err != nil {
 		t.Fatalf("Put: %v", err)
 	}
-	if _, err := kv.Put(ctx, string(event8PutFooBaz.Kv.Key), string(event8PutFooBaz.Kv.Value)); err != nil {
+	if _, err := kv.Put(ctx, string(rev8PutFooBaz.Kv.Key), string(rev8PutFooBaz.Kv.Value)); err != nil {
 		t.Fatalf("Put: %v", err)
 	}
-	if _, err := kv.Put(ctx, string(event9PutFooYoo.Kv.Key), string(event9PutFooYoo.Kv.Value)); err != nil {
+	if _, err := kv.Put(ctx, string(rev9PutFooYoo.Kv.Key), string(rev9PutFooYoo.Kv.Value)); err != nil {
 		t.Fatalf("Put: %v", err)
 	}
-	if _, err := kv.Put(ctx, string(event10PutZoo.Kv.Key), string(event10PutZoo.Kv.Value)); err != nil {
+	if _, err := kv.Put(ctx, string(rev10PutZoo.Kv.Key), string(rev10PutZoo.Kv.Value)); err != nil {
 		t.Fatalf("Put: %v", err)
 	}
+
+	t.Log("Open watches")
+	watches := make([]clientv3.WatchChan, len(tcs))
+	for i, tc := range tcs {
+		watches[i] = watcher.Watch(ctx, tc.key, tc.opts...)
```

Comment:
FYI, you might need to add a sleep here to ensure that next event for watches opened on `rev:0` will be `rev:11`.

Reply:
Maybe put the sleep after the for loop.